### ISSUE TITLE
[HTTP client] Deadlock when using non keep alive connections

### ISSUE THREAD

**Issue by gsnewmark** [ztellman/aleph#282]
In case all requests are sent with `Connection: Close` header (or more accurately if server returns close for all the requests) and number of request are bigger than maximum number of connections in pool client will encounter a deadlock. You can find the minimal reproduction case [here](https://gist.github.com/OlegTheCat/1c89365ed0a61fbaff5cb3a22eff1676).

After the initial investigation I think the culprit is [this `dispose`](https://github.com/ztellman/aleph/blob/5457870cdba22bc3f30891a3b95f3e484b9d8a42/src/aleph/http.clj#L252). Dirigiste's implementation of `dispose` tries to [acquire the lock](https://github.com/ztellman/dirigiste/blob/4f576bc28fe9c18dd43bec59855af4d164558fda/src/io/aleph/dirigiste/Pool.java#L453), but the same lock is already held by other request's try to [acquire the connection](https://github.com/ztellman/dirigiste/blob/4f576bc28fe9c18dd43bec59855af4d164558fda/src/io/aleph/dirigiste/Pool.java#L155). I'm not sure why [timeout](https://github.com/ztellman/aleph/blob/5457870cdba22bc3f30891a3b95f3e484b9d8a42/src/aleph/http.clj#L211) doesn't help.

This problem looks like incorrect usage of pool to me, but if you think issue is with Dirigiste itself I'll move the issue to the project's tracker.

**Comment by ztellman** [comment]
Hmm.  Both `acquire` and `release` shouldn't ever block, they just do a little bookkeeping and release the lock.  Can you give me a stack dump?

**Comment by gsnewmark** [comment]
@OlegTheCat

**Comment by OlegTheCat** [comment]
@ztellman, [here](https://gist.github.com/OlegTheCat/bdaff903df72ce5520d5d539863ffdb7)

**Comment by ztellman** [comment]
I see a single blocked thread at https://gist.github.com/OlegTheCat/bdaff903df72ce5520d5d539863ffdb7#file-dump-txt-L428, which is blocking on a unrealized deferred.  I don't know what that deferred represents, so it's hard to say more.  Can you give me some more context?

**Comment by gsnewmark** [comment]
It's deferred waiting for the completion of all requests https://gist.github.com/OlegTheCat/1c89365ed0a61fbaff5cb3a22eff1676#file-core-clj-L38

**Comment by ztellman** [comment]
Okay, so this doesn't look like a deadlock so much as one or more requests which never completed, unless I'm misunderstanding the stack dump.  Since this is just Aleph talking to Aleph, this still seems like a bug of some sort, but I want to make sure I'm not misunderstanding the problem you're reporting.  Does that seem accurate?

**Comment by gsnewmark** [comment]
Yes, the real problem is that some of requests are waiting for the results indefinitely when keep-alive is disabled. Changing the client pool's [`:keep-alive?` config](https://gist.github.com/OlegTheCat/1c89365ed0a61fbaff5cb3a22eff1676#file-core-clj-L19) doesn't have any visible effect. We've tried both latest stable Aleph (`0.4.1`) and alpha one (`0.4.2-alpha8`).

Deadlock was the first guess, sorry for confusing you. Should I rename this issue?

**Comment by ztellman** [comment]
No, that's fine, I just wanted to make sure I wasn't going off in the wrong direction.  I'm sorry I haven't looked at this sooner, it's a fairly serious issue.  I'll see what I can do.

**Comment by gsnewmark** [comment]
Hi! I have another idea about possible cause for such behavior. In case started batch of requests (let its size be `n`) is bigger than maximum allowed number of objects in pool (`m`), new connections would be created only for the first `m` requests https://github.com/ztellman/dirigiste/blob/4f576bc28fe9c18dd43bec59855af4d164558fda/src/io/aleph/dirigiste/Pool.java#L265. All other `n - m` request would wait for a free connection by adding callback to the underlying queue's `_takes` deque https://github.com/ztellman/dirigiste/blob/4f576bc28fe9c18dd43bec59855af4d164558fda/src/io/aleph/dirigiste/Pool.java#L185. But there won't be any free connections, because after the first `m` requests finish their connections would be removed from the pool instead of returning to it (which is correct behavior for non keep-alive connections). 

By default Aleph [doesn't have a timeout](https://github.com/ztellman/aleph/blob/master/src/aleph/http.clj#L210-L211) for acquiring connection from the pool, so we've had uncompleted requests waiting for a connection indefinitely (original problem that lead me to think about the deadlock). When I add explicit `:pool-timeout` to the request's options, all requests are finished, some of them with timeout exception. Unfortunately in both cases acquire callbacks would remain in the [`_takes` deque](https://github.com/ztellman/dirigiste/blob/4f576bc28fe9c18dd43bec59855af4d164558fda/src/io/aleph/dirigiste/Pool.java#L15) leading to a memory leak-like situation (it's not a real leak because queue is bounded, but pretty close to it).

In case description is not clear enough, I'll provide an example: pool is configured for 5 connections per host, I start 10 requests with `:keep-alive? false` to same host using this pool simultaneously. Each of them adds a callback to the `_takes` deque, but only 5 of them actually generate a new connection (filling the [`_puts` deque](https://github.com/ztellman/dirigiste/blob/4f576bc28fe9c18dd43bec59855af4d164558fda/src/io/aleph/dirigiste/Pool.java#L16) with 5 connections). After this 5 requests are finished due to `:keep-alive? false` their underlying connection are removed from the pool, but no new connections are created, so the remaining 5 callbacks in `_takes` queue will hang there forever. In case I'll later make another request using the same pool, new connection will be generated (as pool is no longer full), one of the previously started requests would finish, but the new one would wait forever instead of it (i. e., callbacks deque behaves like a sliding window with a minumum size of `n - m` or 5 in this example). 

So, in case there're regular bursts of requests exceeding maximum capacity of pool, underlying acquire callbacks deque would slowly but steadily grow up to the maximum allowed size (which is [quite high by default](https://github.com/ztellman/aleph/blob/master/src/aleph/flow.clj#L50)).

One of the possible solutions could be creating new objects during the `dispose` in case there are still some pending acquires and pool isn't full, but at first glance it looks like a hack.

**Comment by ztellman** [comment]
I've discovered the issue, which was in Dirigiste.  When an object was disposed in the pool, it didn't check if there were pending objects, and generate a new object to replace the old one.  This would cause the pending acquires to wait forever.

I've released a new version of Dirigiste `0.1.5-alpha1`, and will cut matching versions of Manifold and Aleph once Sonatype actually makes the Dirigiste jar available.  Thank you for the report, and sorry that it took me this long to resolve the issue.
